### ISSUE TITLE
Underwater artifacts / draw failure on older machines (BIRP)

### ISSUE THREAD

**Issue by Revmatek** [wave-harmonic/crest#1097]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Have you checked the documentation to resolve your problem?

- [X] I have read the linked documentation and could not resolve my issue.

### Current Behavior

Observed behavior: 

![driver_or_shader_error](https://user-images.githubusercontent.com/114962753/202925685-fb3f3b80-3ac9-4bc7-89f4-3a26462c230b.png)

Underwater rendering does not seem to work properly on older machines (specific cards / DX see info at the bottom). I acknowledge upfront that supporting this may be seen as unreasonable or outdated, etc. Just bringing attention to it and shedding some light on my findings. I have spent many hours debugging / troubleshooting this. Maybe there is an easy fix 🤞

### Expected Behavior

Underwater environment should render.

### Steps To Reproduce

1. Download the latest version of Crest from github 
2. Run a build with the Examples.unity added as the default
3. Transfer the build onto a machine with either a Kepler or Fermi card (see description below) or a DX9,10,12
4. Run the exe 
5. Click the next Arrow
6. Observe missing underwater sphere bits
7. Observe line based artifacts that appear to be trying to draw the env
8. Repeat step 5-7
9. Observe this is most apparent on "Underwater lighting"

### Unity Version

2021.3.11.1 (LTS)

### Crest Version

4.15.2

### Render Pipeline

Built-In

### Editor or Standalone

Standalone

### Environment

```markdown
- OS: Windows 7 / 10 / 11
- GPU: GTX 560 - Driver 391.35 / GTX 780M - Driver 425.31
- Graphics API: ~ DX11 / 11_0 / 
- Target Platform: PC
```

### Anything else?

Here is a list of known legacy GPUs that Nvidia no longer updates the drivers for (there are probably more):
https://nvidia.custhelp.com/app/answers/detail/a_id/4656/~/list-of-fermi-series-geforce-gpus
https://nvidia.custhelp.com/app/answers/detail/a_id/4780/~/list-of-kepler-series-geforce-notebook-gpus

I have only confirmed this behavior on 2 GPUs, but to me that is an alarming ratio given the amount of other things that do work, it would no doubt prompt things like "X works fine on my computer, why doesn't this" and either I or players going down rabbit holes trying to figure out the problem, which I would like to avoid if possible. It seems like it will happen on anything that has not had drivers updates after 2019, the exact API / reason is unknown. This was the only common factor I could nail down on my older test machines which I use to make sure things work for folks with older PCs or who have no upgraded for whatever reason. When I found out that driver updates are no longer shipped out, I started leaning towards DX as the culprit, I read some similar issues a while back that were closed because the author simply updated the drivers. It is also worth noting that similar behavior occurs in DX9, DX10, and DX12. I assume this officially only supports DirectX 11. I know that there are lots of people that still use older stuff like this (maybe 10%), but that is potentially a lot of pissed off people that won't really understand why it doesn't work. If it is something that is an easy fix such as adding API calls for different versions that would be much appreciated. I did not dive into the shader to see what was going on, many assumptions on my part / educated guessing on what the problem is based on the machines I have available. Anything with latest DX11 on cards that still receive updates seems to work fine. I understand that this is something that may not be fixable simply because Nvidia no longer supports various older cards. It may be an issue that is fixable for various DX versions. Often times when players have problems with Unity or DX they may attempt to run in a different version to fix it which will not fix this, but might be a nice workaround if any part of this can be fixed. This may also occur with URP/HDRP, untested not currently relevant. This occurs at runtime, but I assume it would also happen if you tried to use the editor on one of these machines as well.

**Comment by daleeidd** [comment]
Thank you. That's a tough one. Did you have any troubles with the underwater in the main.unity scene?

**Comment by Revmatek** [comment]
> Thank you. That's a tough one. Did you have any troubles with the underwater in the main.unity scene?

Here is the result in main, note that you have to modify the scene prior to build and put the camera underwater for it to become apparent.

![error_main](https://user-images.githubusercontent.com/114962753/203156462-a561c411-02a3-4dd9-9792-75c233cddd84.png)

I was just using the example to make the steps to reproduce. I don't think it works in any scene when this occurs. I am happy to try out any blind fixes if they become available, I know this may be difficult to test / verify / troubleshoot.

**Comment by daleeidd** [comment]
Can you try enabling _Underwater Renderer > Debug > Disable Artifact Correction_ and see if that makes a difference? If that doesn't work enable _Underwater Renderer > Debug > View Ocean Mask_ and grab a screenshot of that.

**Comment by Revmatek** [comment]
> Underwater Renderer > Debug > Disable Artifact Correction

This looks like it could lead to a promising workaround / fix. 😁 No clear what visuals / functionality I would lose if I had to avoid it. Will need to test more thoroughly, no doubt hints at the issue:
 
![workaround_disable_artifact_correction](https://user-images.githubusercontent.com/114962753/203316657-d151d4df-6e15-4631-844e-31a87ba033f2.png)


> Underwater Renderer > Debug > View Ocean Mask

I capture this still just in case it proves useful:

![view_ocean_mask](https://user-images.githubusercontent.com/114962753/203316669-2c8148b8-5498-4fe9-ae64-bbc2a12d045e.png)

**Comment by Revmatek** [comment]
> What is your Ocean Renderer > LOD Count?
7 (I think that is the default)

**Comment by Revmatek** [comment]
I am not totally sure those single lines are appearing in the latest version with artifacts disabled depicted in the screenshot. I ran the main scene with latest and flew around and did not see the line artifacts as I do in my current version (which I am still working towards upgrading and they are setup in a different way). All I can say for certain now is that there are issues between 11.0 and 11.1. I do see tiny white dots artifacts which are difficult to capture in a screenshot on latest with artifacts disabled on 11.0.

**Comment by Revmatek** [comment]
Here is some very useful info 🤗:

You can reproduce this problem with the following CLI:
`.\TestLatestCrest.exe -force-d3d11 -force-feature-level-11-0`

Obviously change the exe name to whatever you want. The dot/minor artifacts should be visible on any machine. I think this rules out specific hardware and calls out the DirectX version 11.0 specifically ruling out a lot of the initial speculation I provided.

**Comment by daleeidd** [comment]
Those dots below the horizon are unexpected, but I couldn't reproduce. Any artefacts above the horizon are generally expected with correction disabled.

**Comment by Revmatek** [comment]
Some updated info on this.

I spent some additional hours playing around with `CrestFillMaskArtefacts.compute` and learned a bit about how the mask works with the visualization on the `UnderwaterRenderer.cs`. I was able to develop a complete workaround via cs script that disables artifact correction when < DX11.1 is detected (i.e. basically just 11.0), the script also lowers this:
```
        [Tooltip("Adjusts the far plane for horizon line calculation. Helps with horizon line issue.")]
        public float _farPlaneMultiplier = 0.68f;
```
to around 0.245 in my case. I did not see any drawbacks to this on either new or old machines, nor artifacts from disabling so for the moment the priority of this for me has dropped to near zero, unless my hacky workaround revels other issues down the road. The workaround seems to work for both Keplar and Fermi series cards so I assume this fixes the issue for Nvidia at least across the board.

To my dismay I was unable to find a fix for this bug. But I did gain some additional hints at what the problem may be at a lower level. The main thing I noticed was that there definitely seems to be some sort of race condition with the compute shader (perhaps in DX11.0 itself or the drivers...). I see that ids are being used for the groups but it seems like it may be possible somehow that pixels are being overwritten when they match and then that value >0 (above water?) or something is being copied across the entire screen above and below water. 
```
[numthreads(8, 8, 1)]
```
influences this behavior greatly. Reducing the number of threads seems to reduce the issue while not resolving it completely.

Thanks for the help!
